### ISSUE TITLE
New component: ListboxControl

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#65801]
> A listbox widget presents a list of options and allows a user to select one or more of them. A listbox that allows a single option to be chosen is a single-select listbox; one that allows multiple options to be selected is a multi-select listbox.
> – [w3.org](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/)

Listboxes can be used as controls in their own right, and also power more complex components like Comboboxes and Autocomplete inputs.

### Listbox Item

<img width="1892" alt="Listbox Item" src="https://github.com/user-attachments/assets/c077bb71-5094-4c67-882f-69294115adb1">

This is the atomic element that makes up the majority of the Listbox component. It can appear as either a checkbox for multi-select listboxes or a radio for single-select listboxes. In the multi-select variant the checkbox is not a checkbox input, merely an element styled to look like one.

### Listbox

<img width="1892" alt="Listbox" src="https://github.com/user-attachments/assets/e64b2e1a-32ff-4457-a980-84f46bdc743e">

Listbox is comprised of several elements; notably:

* A scrollable container
* Listbox Items
* Listbox Item groups
* Listbox group headers

### ListboxControl

<img width="1892" alt="ListboxControl" src="https://github.com/user-attachments/assets/99644e88-abe2-473a-9746-98f82895a8a8">

ListboxControl is the consumable component. It follows the same format as similar components by providing layout and optional elements like a label and help text. Here's a video demonstration of the assembled component:

https://github.com/user-attachments/assets/8de50935-473a-4a6e-a9af-edbf3928b2b6

**Comment by ciampo** [comment]
Thank you for working on this!

In the case of [listbox widgets](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/), there is no such thing as "checkbox" or "radio" items . Rather, there is only one type of listbox options — and there can be none, one, or multiple selected at the same time.

To avoid the confusion with radios and checkboxes, I wonder if we should consider a different visual indicator for the selected items (for example, [ariakit uses a tick](https://ariakit.org/examples/combobox-multiple) in their multiple selection combobox example)

cc @WordPress/gutenberg-components

**Comment by DaniGuardiola** [comment]
I wonder if we could conceptualize this as a "MenuList", as that'd allow us to, instead of creating a new component, make the existing menu component cover more use cases by making it more open. This mimic the Ariakit structure (see [`MenuList`](https://ariakit.org/reference/menu-list)) and, since we already use it for the menu, it'd make this easier to implement. Overall, I think it'd greatly lower the design, maintenance, education, and support overhead. It'd just be one more Storybook story in that component.

To be clear, what I am proposing is to add the possibility to use our current menu (`DropdownMenuV2`) without a popover or a trigger. This is already supported by Ariakit and valid in terms of a11y, and the design looks pretty similar to the screencaps in this issue description:

<img width="302" alt="image" src="https://github.com/user-attachments/assets/fdb8cb77-6cc5-4bcb-93d1-af15b5bb322f">

<img width="202" alt="image" src="https://github.com/user-attachments/assets/085dfd1e-996a-466e-99e0-0bdb200db6db">


@ciampo's suggestion of using a tick is also covered by this.

**Comment by ciampo** [comment]
[Listboxes](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/) are more generic and can be the foundation of components like `CustomSelectControl`, `ComboboxControl` and `FormTokenField` — [menus](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/) are more specific and have their different roles and semantics. I'm not sure they are compatible to the point of sharing the same underlying implementation.

We could share a lot of the styles, though.

**Comment by hanneslsm** [comment]
I like the design of both the checkbox and the radio, but - as pointed out by @ciampo and @DaniGuardiola - they both look like there are from different design languages.
Either the Radio is missing a border or the Checkbox shouldn't have one as in the DropdownMenu2
UX wise, I'd go with a design that shows the border even when not selected. At least on hover.

Another thing is the color blue. I like it. 
But currently, most components use black as indicator for being selected.

**Comment by hanneslsm** [comment]
Maybe we also should consider ~"half ticked" (or how would you say that?)~ the [indeterminate boolean property](https://css-tricks.com/indeterminate-checkboxes/)
<img width="60" alt="CleanShot 2024-10-02 at 12 54 37@2x" src="https://github.com/user-attachments/assets/9380cb1c-bacc-4f5d-a4de-109d6c73c4ad"> 
<img width="247" alt="CleanShot 2024-10-02 at 13 00 05@2x" src="https://github.com/user-attachments/assets/4842b020-4883-48e4-b1f6-171850a99ce7">

**Comment by ciampo** [comment]
> Maybe we also should consider ~"half ticked" (or how would you say that?)~ the [indeterminate boolean property](https://css-tricks.com/indeterminate-checkboxes/) 

As mentioned above, if we're sticking with `listbox` semantics, there is no such thing as checkbox semantics, and therefore I don't think that indeterminate state would apply.

At most, that could be added to places where we support `menu` semantics (like `DrodownMenu` and `DropdownMenuV2`), since those semantics support [`menuitemcheckbox` roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menuitemcheckbox_role)

**Comment by DaniGuardiola** [comment]
On my phone right now so can't check, but I think it's worth looking into the semantics that Ariakit has for MenuList. Wouldn't surprise me if it just used plain listbox semantics. The name of Ariakit components does not always correlate to the corresponding semantics.

**Comment by jameskoster** [comment]
> there is no such thing as "checkbox" or "radio" items

@ciampo I appreciate they are not semantically checkbox or radio items, but isn't it correct that in terms of behavior listboxes can be configured either as single-select (similar to a `select`) or multi-select (kind of like a group of checkboxes). Different style treatment can help users identify the behavior, even if it's not semantically accurate.

I suppose I should have noted that the design would not utilise _actual_ checkboxes, they would be elements styled to look like checkboxes. This is a fairly common pattern in design systems; there's a similar example in github issue label assignment UI:

<img src="https://github.com/user-attachments/assets/9e5bd100-f1cc-4167-a3cd-05de496d5ed2" width="300">

**Comment by mirka** [comment]
I'm trying to think what the usage guidelines for this would look like. Specifically, when should one use this Listbox over other similar controls, like:

## Single-select

- SelectControl
- ComboboxControl (note that in the v2, we'll be able to add groupings/headings)
- RadioControl

## Multi-select

- CustomSelectControlV2 multi-select mode
- FormTokenField
- CheckboxControl

I'm not sure you'd need a separate Listbox component unless it is for a feature that is not covered in the alternatives, like making the items sortable (as in #64686). And if the main feature we want to provide is sortability, then it might make more sense to provide it as a lower-level, headless API (e.g. [SortableJS](https://sortablejs.github.io/Sortable/)) so consumers can add sortability to a broader variety of designs.

**Comment by ciampo** [comment]
> I think it's worth looking into the semantics that Ariakit has for MenuList. Wouldn't surprise me if it just used plain listbox semantics.

@DaniGuardiola I believe that they apply `role="menu"`. But even if that wasn't the case, having a checkbox or a radio item just doesn't make sense in a `listbox` widget at the standard level, right?

> I appreciate they are not semantically checkbox or radio items, but isn't it correct that in terms of behavior listboxes can be configured either as single-select (similar to a `select`) or multi-select (kind of like a group of checkboxes). Different style treatment can help users identify the behavior, even if it's not semantically accurate.
> 
> I suppose I should have noted that the design would not utilise _actual_ checkboxes, they would be elements styled to look like checkboxes. This is a fairly common pattern in design systems; there's a similar example in github issue label assignment UI:

@jameskoster I see your point — let me try to rephrase my feedback:

"radio" and "checkbox" are concepts that don't apply in the context of a listbox, and I'm afraid that using them even as just visual cues may set the wrong expectation to the user.

Basically, we should just pick one style (probably checkbox-like), instead of using both checkbox-like and radio-like

> I'm trying to think what the usage guidelines for this would look like

@mirka I am seeing this component as a very low-level component that we can leverage to implement all of the other components that you mentioned? Similarly to Composite`

**Comment by jameskoster** [comment]
> Basically, we should just pick one style (probably checkbox-like), instead of using both checkbox-like and radio-like

Don't you think there's value in the user understanding at a glance whether they can select one or many options? For example consider the the Status filter in the Pages data view. Currently the UI gives no indication whether the list is multi-select or single-select which leaves questions about the behavior:

<img width="350" src="https://github.com/user-attachments/assets/2425bdbf-cd3a-4de1-a96f-225bb1fb478f">

If you see empty checkboxes then it's much clearer that you can select multiple options.

**Comment by mirka** [comment]
> I am seeing this component as a very low-level component that we can leverage to implement all of the other components that you mentioned?

Of the components I listed, those that could possibly use this is ComboboxControl/FormTokenField (which will eventually be a unified as ComboboxControlV2) and CustomSelectControlV2. Ariakit Select and Combobox already have their own subcomponents for these things. So in that case, it would be more natural to just share/reuse the styles, not entire components, because Ariakit is already providing them (behavior + semantics).

The only reason we would want a standalone Listbox component (styles + behavior + semantics) is if we want to use it independently, or in a non-Ariakit-based component. And I'm struggling to imagine such a use case (aside from the sortable case).

**Comment by ciampo** [comment]
> The only reason we would want a standalone Listbox component (styles + behavior + semantics) is if we want to use it independently, or in a non-Ariakit-based component. And I'm struggling to imagine such a use case (aside from the sortable case).

The example that Jat posted [above](https://github.com/WordPress/gutenberg/issues/65801#issuecomment-2388974037) ( the dataviews filter) could be such a use case?

**Comment by mirka** [comment]
> The example that Jat posted [above](https://github.com/WordPress/gutenberg/issues/65801#issuecomment-2388974037) ( the dataviews filter) could be such a use case?

I think I would prefer just making CustomSelectControlV2 flexible enough to support that design. Right now, `renderSelectedValue` doesn't allow that level of customization, but we should if design requires it.

**Comment by ciampo** [comment]
> > The example that Jat posted [above](https://github.com/WordPress/gutenberg/issues/65801#issuecomment-2388974037) ( the dataviews filter) could be such a use case?
> 
> I think I would prefer just making CustomSelectControlV2 flexible enough to support that design. Right now, `renderSelectedValue` doesn't allow that level of customization, but we should if design requires it.

Definitely something to consider. We'd also need to make sure we can offer the listbox itself separate from the popover / dropdown part, though, because this UI pattern can also happen:

![Image](https://github.com/user-attachments/assets/698788a2-ab80-4f77-996a-90019dd6eb05)

**Comment by jarekmorawski** [comment]
Great discussion! I'm so glad we're talking about changes to the select component because it's likely the second most-used component in the admin experience (including screens added by plugins). I don't want to stir the pot too much, but I'd love to see us have just one select component with all the necessary features.

Here's my wishlist based on my experience designing admin pages. Most of these features focus on the multi-select variant, but some would also apply to the single-select version.

---

### Support for nested groups

There are many cases where users need to browse items organized in hierarchies. The component should support displaying items not just in groups with individual headings ([like in Jay's example](https://github.com/WordPress/gutenberg/issues/65801#issue-2559655097)), but also in a tree-like view. This would allow users to browse items contained within other items. These groups could be displayed as accordions, enabling users to expand and collapse groups, and to select them with a single click. 

- **If a parent is selected**, all items inside it are selected.  
- **If only some items inside are selected**, the parent is displayed in an intermediate state.  
- **Clicking again** would select or deselect all items inside.

#### Use cases:

- Adding categories and subcategories when publishing a post
- Choosing supported languages in the WPML plugin
- Selecting accepted currencies in WooCommerce

![Image](https://github.com/user-attachments/assets/108edf96-a58b-413c-b258-32902f6d65b5)

---

### "Contained" display mode

I agree that the `ListboxControl` component should work independently of a dropdown or select component. The container should mimic the styling of a standard select field, including border size, color, and radius.

#### Use cases:

- Adding a new page and choosing its placement in the menu structure
- Adding a new category and choosing its parent category

![Image](https://github.com/user-attachments/assets/6db1552f-bcb3-4d8b-b512-17a09c27e470)

---

### Support for select/deselect all

It would be helpful to allow users to select or deselect all items in the list. Selecting all could be done with a dedicated button at the top of the list (not fixed), while clearing the selection could be handled via an "X" button on the right-hand side of the field. This is a standard feature in many established design systems.

#### Use cases:

- Choosing fields in a data sync plugin
- Selecting recommended linked products in WooCommerce

![Image](https://github.com/user-attachments/assets/36130cc3-7ee3-4feb-ac23-a25fc64aa821)

---

### Display selected items as tags

Currently, the multi-select variant of the `CustomSelectControl` v2 component provides limited feedback to users about their selections.

![Image](https://github.com/user-attachments/assets/7d4c128a-5aa9-49b1-a432-d1759b180ba3)

Could we display selected items as tags instead and merge this component with [FormTokenField](https://wordpress.github.io/gutenberg/?path=/docs/components-formtokenfield--docs)? Both serve similar purposes, and showing selections as tags would make it easier for users to reorder or remove items.

#### Use cases:

- Adding and reordering tags in a post
- Adding products displayed in WooCommerce blocks
- Managing individual filter values in a data table

![Image](https://github.com/user-attachments/assets/72833b6b-5f0b-4f7b-bd2b-510726512b66)

---

### Support creating new items

Select components often allow users to create new items directly. While we currently support this functionality in some editor workflows using separate buttons, would it be feasible to integrate it into the listbox component? 

For instance, a "create new item" option could appear fixed at the bottom of the listbox, opening a modal when activated by clicking or pressing `Enter`. This feature could also work when searching, allowing users to create new items with the search term set as the default label in the creation interface.

#### Use cases:

- Assigning a post to new categories
- Adding tags to a product in WooCommerce

![Image](https://github.com/user-attachments/assets/e6058b56-ed96-41b7-9a55-b96470aad6f2)

Hope this helps! cc @brookewp @WordPress/gutenberg-design #66173

**Comment by ciampo** [comment]
Hey @jarekmorawski , thank you for sharing your feedback!

Here are some thoughts after reading your feedback — looking forward to more feedback from the rest of the folks involved

> I don't want to stir the pot too much, but I'd love to see us have just one select component with all the necessary features.

I'm not sure that's a good approach, as it would incur the risk of blurring the lines too much and mering components and widgets that are semantically different.

While, for example, I agree that we could look at "merging" `FormTokenField` into `ComboboxControl`, I don't think that `SelectControl` and `ComboboxControl` should be merged under the same component.

I'm also noticing that every example design in your list is showing a combobox, instead of a listbox. A [listbox](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/) can be part of a combobox widget, but a [combobox widget](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/) could also use a `grid`, `tree` or `dialog` instead of a `listbox` for its popup.

> Support for nested groups

What you're describing here is a [Tree View](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/) — and in your example, you're using the treeview as part of a `combobox` widget. Given than a tree view is a separate widget, with different semantics and affordances from a listbox, I suggest that we look at it and implement it as a separate component. And only if we find strong similarities with `listbox`, we consider merging the functionalities with the right APIs to differentiate between the two.

As of today, the `@wordpress/components` package offers the `TreeSelect` component, but at a quick glance it doesn't support all the required features (and I'd need a deeper look to understand if it's worth improving the existing component vs creating a new one from scratch).

I will also note that building such component can be tricky, especially in terms of adherence to the WAI-aria specs and the accessibility requirements, and that as of today, most libraries (including [ariakit](https://github.com/ariakit/ariakit/issues/4059#issuecomment-2527893226)) don't offer a `TreeView` base component.

> "Contained" display mode

I agree with you on this one. Exposing an "inline" list as an alternative to the "popover-based" version is something that we intend to do with a few components (`Menu`, `SelectControl`, etc..)

> Support for select/deselect all

- Regarding a "select all" item as the first item in the list, we'd need to understand if it can be done in a semantically valid way, since technically that wouldn't be a valid option item, but more of a button.
- The "deselect all" button shouldn't be an issue for the combobox case. In the "pure" select or tree view case it may need to be a separate button, since the dropdown toggle is already a button in itself;

Since you mentioned that "This is a standard feature in many established design systems", it would be great if you could point to UI libraries or products implementing these patterns, which can help us find inspiration on how to implement it on our side.

> Display selected items as tags

- The `CustomSelectControlV2` component is still at its very inception (and not actively being worked on). We don't recommend folks to use it in the code or even as a reference for the time being;
- In any case, as also explained above, it wouldn't make sense to merge `FormTokenField` into `CustomSelectControl`, given how they implement different underlying semantics (`listbox` vs `combobox`);
- Having said that, we could still look alternative ways to represent the selected item(s) when the popover is closed. Although, as explained above, `CustomSelectControl`'s popover trigger is a  button, and as such it can't have interactive elements as children components (such as individually selectable and deletable tags).

> Support creating new items

- Your example shows a combobox widget — which is probably the best-suited widget for this UX, since it already displays an input field to type the new item. However, I can still think of a few challenges:
  - while the combobox widget natively allows the user to input a value that doesn't match any suggestion, I'm not sure it is ok to add a new "option" to the list — and if that was the case, we'd probably need to communicate this change to assistive technology users;
  - for non-combobox widgets (eg. listbox), we'd have to display an additional text field for the user to enter the new item — which, again, feels quite bespoke and something we'd need to make sure we implement correctly for assistive technology users;
- This poses a similar challenge to the "Select all" item: we'd need to find a way to render an additional button that is not actually part of the list of items in the popover, without breaking the semantics. We'd need to determine keyboard interactions and other edge case behaviors too (for example, should / can the "Add new item" button be part of the same composite widget as the items?)

**Comment by jarekmorawski** [comment]
Thank you so much for a detailed response, @ciampo. The semantics are essential and I'm sure we can work out a good solution. I now see that merging all components may not be the way forward and that's perfectly fine as long as the different components we create share DNA, are linked together, and it's clear to consumers when and how they should be used.

> As of today, the @wordpress/components package offers the TreeSelect component, but at a quick glance it doesn't support all the required features (and I'd need a deeper look to understand if it's worth improving the existing component vs creating a new one from scratch).

It is one of the reasons why we built a custom tree select component at WooCommerce. I believe now is our best chance to improve the one that already exists.

> I will also note that building such component can be tricky, especially in terms of adherence to the WAI-aria specs and the accessibility requirements, and that as of today, most libraries (including https://github.com/ariakit/ariakit/issues/4059#issuecomment-2527893226) don't offer a TreeView base component.

That is unfortunate, but I really hope we can find an approach that strikes a good balance between usability and accessibility. There are several important flows in WooCommerce and other products I work on that would benefit from a robust tree select component. Creating a version of it that meets all the accessibility requirements at the expense of speed and efficiency would be suboptimal for complex flows in Woo where every click matters and either takes you closer or further away from your goal.

> Since you mentioned that "This is a standard feature in many established design systems", it would be great if you could point to UI libraries or products implementing these patterns, which can help us find inspiration on how to implement it on our side.

Here are examples from [Carbon (IBM)](https://carbondesignsystem.com/components/dropdown/usage/#multiselect), [Atlassian](https://atlassian.design/components/select/checkbox-select/examples), [Helsinki](https://atlassian.design/components/select/checkbox-select/examples), and [Elvia](https://design.elvia.io/components/dropdown?isMulti=true&hasSelectAllOption=true). All of them also feature other improvements I'm pitching, like nesting and grouping. Atlassian's select component comes with a `creatable` flag.

> I'm not sure it is ok to add a new "option" to the list

Here's an example of a [potentially accessible implementation](https://react-select.com/creatable).

> we'd need to find a way to render an additional button that is not actually part of the list of items in the popover, without breaking the semantics.

Not sure if it helps, but most implementations I reviewed used a group parent item to replicate the functionality of a select/deselect button. Because it's a regular list item with custom functionality, it didn't require any changes to the semantics. Perhaps the 'Create' button could work similarly, or appear as a single button only when there are no results in the list (in such a case, we'd no longer mix components because it wouldn't be part of the same widget).

**Comment by jarekmorawski** [comment]
[Example of a combobox](https://www.lightningdesignsystem.com/components/combobox/#Open-with-Two-or-More-Options-Selected) where the listbox is combined with an additional footer containing contextual buttons.

![Image](https://github.com/user-attachments/assets/627e63a1-2e8b-4d51-947a-cb283d5796f8)

**Comment by ciampo** [comment]
Thank you for sharing more examples! Here are a few more notes:

- from what I can see, the "Select all" and "Create new item" functionality are done by adding an extra `option` item in the list — which fixes the "semantics" part, although it feels a bit hacky. We'll need to test the approach and see if we can make it work well;
- While the "select all" option makes sense both in a listbox and in a combobox, the "add new item" seems to be a feature only implemented in comboboxes (since, as mentioned before, they already have an input field)
- Few examples implement a "clear all" button:
  - I haven't found any doing so on with a `listbox` widget (likely due to the reasons I explained above)
  - Some do with a combobox (for example, [Ariakit](https://ariakit.org/examples/combobox-cancel)), although there are concerns in terms of accessibility and keyboard navigation (see the "Remove clear control" section [here](https://atlassian.design/components/select/usage#accessibility)). For example, MUI's [combobox](https://mui.com/material-ui/react-autocomplete/#combo-box) excluded the clear button from tab order.
- I couldn't find code examples of a nested, collapsible tree view in the links that you posted (please point me to the right examples in case I missed them!). I took some time to research existing implementation, and here's what I found:
   - MUI has a [tree view component](https://mui.com/x/react-tree-view/) as part of their premium (paid) MUI x suite of components. The closest version to what we're discussing here is [this example](https://ariakit.org/examples/combobox-cancel) (note that some of the features are still in beta)
   - React aria has a [Tree component](https://react-spectrum.adobe.com/react-aria/Tree.html), but it's still in beta and quite incomplete:

**Comment by jarekmorawski** [comment]
> While the "select all" option makes sense both in a listbox and in a combobox, the "add new item" seems to be a feature only implemented in comboboxes (since, as mentioned before, they already have an input field)

That's true and I think a combobox can be a good start. It's my primary focus as most data in the admin is text or numbers. I'm curious about the other widget types you mentioned. Do we already use them somewhere?

> I haven't found any doing so on with a listbox widget (likely due to the reasons I explained above)

[Salesforce's design system](https://www.lightningdesignsystem.com/components/combobox/#Selecting-an-option-3) replaces the input's contents with the selected item. It looks like they're stretching the `pill` component to fill the container, which isn't great, but does the job, I suppose.

Displaying the X button next to the chevron could work for single-select items ([example from an existing DS](https://docs.tetrisly.com/components/list/select), [another one](https://eui.elastic.co/#/forms/combo-box#single-selection)).

![Image](https://github.com/user-attachments/assets/d329b65e-bd2c-4f74-9543-1045ac4c5103)

> please point me to the right examples in case I missed them!

Here's [one](https://ant.design/components/tree-select) and [another](https://www.telerik.com/design-system/docs/components/dropdowntree/). Not sure they're great references, though.

**Comment by jameskoster** [comment]
Noting that `listbox` may not be appropriate for category selection given the need to visualise structure. A [`Treeview`](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/) might work better. 

Both `listbox` and `treeview` can benefit from being combined with an input, in the case of the former to create a `combobox`, and in the latter to create something like a `combotree`. There's a good example of that here: [TreeSelect](https://ant.design/components/tree-select).
